[4.1] Child templates 2/2 [clean]

dgrammatiko opened the pull request:
Pull Request for Issue # .


### Summary of Changes
- Added functionality for the creation of a child template (for templates that support it). The name convention for the child template is [parent template]_[child template name] (eg for cassiopeia's child named temp the actual name will be `cassiopeia_temp`
- Tweaked the language inclusions so children templates of the core get a correct translation (templates that have the language folder were already covered)
- Tweaked the functionality for the preview thumb and preview images. Also added a fallback image if there's not an image in the template...
- Tweaked the file editor so that folders could be created/deleted/renamed both on the template folder or the media folder
- Tweaked the file editor so that files could be edited/created/deleted/renamed/uploaded/extracted both on the template folder or the media folder

### Testing Instructions
- Git fetch this PR, run npm
- Try every functionality in the com_templates for both core templates

- Install https://github.com/joomla/joomla-cms/files/7235154/cassy_1.0.0.zip (cassiopeia with a different name)
- Try every functionality in the com_templates for the installed template

- Also check (with tinymce as the editor) when editing an article that the editor.css (or .min.css if not in debug mode) is loaded from the `/css` folder of the template, if such file exists. (the path can be `/templates/templateName/css/editor.[min.]CSS` for existing templates or  `media/templates/site/templateName/css/editor.[min.]CSS` for templates supporting children).

### Actual result BEFORE applying this Pull Request



### Expected result AFTER applying this Pull Request



### Documentation Changes Required

Yes, this is the UI for the codename `child template` or whatever the project ends up naming the inheritable/extendable/... templates

dgrammatiko committed:
Fix the missing installer info

dgrammatiko committed:
Fix the template thumb images

dgrammatiko committed:
Fix missing language on style view

brianteeman commented on administrator/components/com_templates/src/Service/HTML/Templates.php:
```diff
@@ -59,38 +82,66 @@ public function thumb($template, $clientId = 0)
 	 * @return  string  The html string
 	 *
-	 * @since   3.4
+	 * @since   1.6
```
copy paste error?

dgrammatiko replied:
Yup. Fixed, thanks

brianteeman commented on build/media_source/system/images/template_thumb.svg:
```diff
@@ -0,0 +1,24 @@
+		<path fill="#020202" d="M403.191,172.262c0.547-0.492,1.089-0.994,1.616-1.521l21.969-21.967c12.998-13,14.935-32.139,4.323-42.75
+			l-33.302-33.301c-10.61-10.611-29.749-8.676-42.748,4.322l-21.968,21.969c-0.528,0.527-1.031,1.068-1.522,1.617L403.191,172.262z"
+			/>
+		<polygon fill="#020202" points="99.842,403.98 135.257,439.396 88.128,451.109 41,462.822 52.712,415.691 64.425,368.563 		"/>
```
```suggestion
		<polygon fill="#020202" points="99.842,403.98 135.257,439.396 88.128,451.109 41,462.822 52.712,415.691 64.425,368.563 "/>

```

dgrammatiko replied:
This is a temporary file (placeholder) till we get the actual image.  @bembelimen already has assigned this task to some people/group afaik

brianteeman commented on build/media_source/system/images/template_thumb.svg:
```diff
@@ -0,0 +1,24 @@
+			/>
+		<polygon fill="#020202" points="99.842,403.98 135.257,439.396 88.128,451.109 41,462.822 52.712,415.691 64.425,368.563 		"/>
+		<polygon fill="#020202" points="340.238,132.844 108.008,365.074 97.761,354.828 329.99,122.598 319.744,112.352 74.961,357.133
+			146.689,428.859 391.471,184.079 		"/>
```
```suggestion
			146.689,428.859 391.471,184.079 "/>

```

brianteeman commented on libraries/src/Installer/Installer.php:
```diff
@@ -2302,6 +2302,17 @@ public static function parseXMLInstallFile($path)
 		$data['description'] = (string) $xml->description;
 		$data['group'] = (string) $xml->group;
 
+		// Child templates specific fields.
```
```suggestion
		// Child template specific fields.
```

dgrammatiko committed:
Update Templates.php

dgrammatiko committed:
Update libraries/src/Installer/Installer.php

Co-authored-by: Brian Teeman <brian@teeman.net>

dgrammatiko committed:
Update build/media_source/system/images/template_thumb.svg

Co-authored-by: Brian Teeman <brian@teeman.net>

dgrammatiko committed:
Update build/media_source/system/images/template_thumb.svg

Co-authored-by: Brian Teeman <brian@teeman.net>

dgrammatiko committed:
Fix the thumb logic

dgrammatiko commented:
@brianteeman do you know if the `aria-setsize`, `aria-posinset` and `aria-level` are fully supported on all evergreen browsers? Ref: https://w3c.github.io/aria-practices/examples/treeview/treeview-1/treeview-1b.html

brianteeman commented:
Sorry I don't know. I will do some research tomorrow. Caveat a lot of the examples on that site are not very good

dgrammatiko commented:
Do you have a solid example for tree view? I use some of the links from https://github.com/w3c/aria/issues/1311

brianteeman commented:
Not at my pc but the link I shared the other day in the media manager pr looked good

dgrammatiko commented:
> Not at my pc but the link I shared the other day in the media manager pr looked good

It's sharing the same code as https://w3c.github.io/aria-practices/examples/treeview/treeview-1/treeview-1b.html
Anyways I will code something and we can improve/iterate...

dgrammatiko committed:
editor #1

dgrammatiko committed:
CS

dgrammatiko committed:
CS

dgrammatiko committed:
CS + fix logic

dgrammatiko committed:
Ensure the isMedia flag is correcly propageted

dgrammatiko committed:
Ooops, forgot the child creation functionality

From #35879

dgrammatiko committed:
CS

brianteeman commented on administrator/components/com_templates/tmpl/template/default_modal_child_body.php:
```diff
@@ -0,0 +1,33 @@
+ * @package     Joomla.Administrator
+ * @subpackage  com_templates
+ *
+ * @copyright   (C) 2016 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright   (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

it is a new file

brianteeman commented on administrator/components/com_templates/tmpl/template/default_modal_child_footer.php:
```diff
@@ -0,0 +1,16 @@
+ * @package     Joomla.Administrator
+ * @subpackage  com_templates
+ *
+ * @copyright   (C) 2016 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright   (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

it is a new file

dgrammatiko committed:
Update administrator/components/com_templates/tmpl/template/default_modal_child_body.php

Co-authored-by: Brian Teeman <brian@teeman.net>

dgrammatiko committed:
Update administrator/components/com_templates/tmpl/template/default_modal_child_footer.php

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented on administrator/components/com_templates/tmpl/template/default_modal_child_body.php:
```diff
@@ -0,0 +1,33 @@
+			<div class="control-group">
+				<div class="control-label">
+					<label for="new_name">
+						<?php echo Text::_('COM_TEMPLATES_TEMPLATE_CHILD_NAME_LABEL'); ?>
+					</label>
+				</div>
+				<div class="controls">
+					<input class="form-control" type="text" id="new_name" name="new_name" required>
+					<small class="form-text">
+						<?php echo Text::_('COM_TEMPLATES_TEMPLATE_CHILD_NAME_LABEL'); ?>
+					</small>
```
Do you really mean to display the same text twice?

dgrammatiko replied:
Nope, the string is wrong

dgrammatiko replied:
Good catch, https://github.com/joomla/joomla-cms/pull/35998/commits/e5a4b0b34224ec903040b9ce829ba60c1d793f6b should fix that

brianteeman replied:
the new string does not exist

dgrammatiko replied:
It should, it's used in the copy modal

dgrammatiko replied:
![Screenshot 2021-11-18 at 15 50 56](https://user-images.githubusercontent.com/3889375/142438380-17a56232-ed40-4958-9100-6441861652c5.png)

dgrammatiko committed:
Fix the description for the child name

brianteeman commented on administrator/components/com_templates/tmpl/template/default_media_folders.php:
```diff
@@ -0,0 +1,34 @@
+ * @package     Joomla.Administrator
+ * @subpackage  com_templates
+ *
+ * @copyright   (C) 2013 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright   (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

it is a new file

brianteeman commented on administrator/components/com_templates/tmpl/template/default_tree_media.php:
```diff
@@ -0,0 +1,73 @@
+ * @package     Joomla.Administrator
+ * @subpackage  com_templates
+ *
+ * @copyright   (C) 2013 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright   (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

it is a new file

dgrammatiko replied:
Thanks

dgrammatiko committed:
Update administrator/components/com_templates/tmpl/template/default_media_folders.php

Co-authored-by: Brian Teeman <brian@teeman.net>

dgrammatiko committed:
Update administrator/components/com_templates/tmpl/template/default_tree_media.php

Co-authored-by: Brian Teeman <brian@teeman.net>

dgrammatiko committed:
Fix copy/pasting doc blocks

dgrammatiko committed:
Child tmpl name === parentTmplName_childName

dgrammatiko committed:
Merge branch '4.1-dev' into 4.1-dev-com_templates_clean

brianteeman commented on administrator/components/com_templates/src/Model/TemplateModel.php:
```diff
@@ -1472,17 +1496,17 @@ public function createFolder($name, $location)
-			$client = ApplicationHelper::getClientInfo($template->client_id);
-			$path   = Path::clean($client->path . '/templates/' . $template->element . '/');
+			$path   = Path::clean($location . '/');
+			$base = $this->getBasePath();
```
```suggestion
			$base   = $this->getBasePath();
```

brianteeman commented on administrator/components/com_templates/src/Model/TemplateModel.php:
```diff
@@ -1985,4 +1989,144 @@ protected function checkFormat($ext)
+		$result = array();
+		$template = $this->getTemplate();
+
+		if (!isset($template->xmldata)) {
+			$template->xmldata = TemplatesHelper::parseXMLTemplateFile($template->client_id === 0 ? JPATH_ROOT : JPATH_ROOT . '/administrator', $template->name);
+		}
```
```suggestion
		if (!isset($template->xmldata)) 
		{
			$template->xmldata = TemplatesHelper::parseXMLTemplateFile($template->client_id === 0 ? JPATH_ROOT : JPATH_ROOT . '/administrator', $template->name);
		}
```

brianteeman commented on administrator/components/com_templates/src/Model/TemplateModel.php:
```diff
@@ -1985,4 +1989,144 @@ protected function checkFormat($ext)
+		$path = Path::clean(JPATH_ROOT . '/media/templates/' . ($template->client_id === 0 ? 'site' : 'administrator') . '/' . $template->element . '/');
+		$this->mediaElement = $path;
+
+		if (!is_writable($path)) {
+			$app->enqueueMessage(Text::_('COM_TEMPLATES_DIRECTORY_NOT_WRITABLE'), 'error');
+		}
```
```suggestion
		if (!is_writable($path)) 
		{
			$app->enqueueMessage(Text::_('COM_TEMPLATES_DIRECTORY_NOT_WRITABLE'), 'error');
		}
```

dgrammatiko replied:
Thanks

dgrammatiko committed:
Update administrator/components/com_templates/src/Model/TemplateModel.php

Co-authored-by: Brian Teeman <brian@teeman.net>

dgrammatiko committed:
Update administrator/components/com_templates/src/Model/TemplateModel.php

Co-authored-by: Brian Teeman <brian@teeman.net>

dgrammatiko committed:
Update administrator/components/com_templates/src/Model/TemplateModel.php

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented on administrator/components/com_templates/src/View/Template/HtmlView.php:
```diff
@@ -146,15 +155,16 @@ class HtmlView extends BaseHtmlView
-		$this->state    = $this->get('State');
-		$this->template = $this->get('Template');
-		$this->preview  = $this->get('Preview');
+		$app              = Factory::getApplication();
```
```suggestion
		$app               = Factory::getApplication();
```

dgrammatiko committed:
Update administrator/components/com_templates/src/View/Template/HtmlView.php

Co-authored-by: Brian Teeman <brian@teeman.net>

dgrammatiko committed:
CS

Quy commented on administrator/components/com_templates/src/Service/HTML/Templates.php:
```diff
@@ -26,31 +27,67 @@ class Templates
+		}
 
-		if (file_exists($thumb))
+		$html   = '';
+		$client = ApplicationHelper::getClientInfo($template->client_id);
```
```suggestion
		$client = ApplicationHelper::getClientInfo($template->client_id);
```

dgrammatiko committed:
Update administrator/components/com_templates/src/Service/HTML/Templates.php

Co-authored-by: Quy <quy@fluxbb.org>

bembelimen committed:
Merge branch '4.1-dev' into 4.1-dev-com_templates_clean

bembelimen commented on administrator/components/com_templates/src/Controller/TemplateController.php:
```diff
@@ -980,4 +984,112 @@ public function ajax()
+		$model->setState('to_path', $app->get('tmp_path') . '/' . $model->getState('tmp_prefix'));
+
+		// Process only if we have a new name entered
+		if (strlen($newName) > 0)
```
Could we convert this to an early return?
```
    if (!strlen($newName))
    {
        $this->setMessage(Text::sprintf('COM_TEMPLATES_ERROR_INVALID_TEMPLATE_NAME'), 'error');

        return false;
    }
```

dgrammatiko replied:
Sure, fwiw this method should be almost identical to the `copy()` one. Will fix this later on today

bembelimen commented on administrator/components/com_templates/src/Controller/TemplateController.php:
```diff
@@ -980,4 +984,112 @@ public function ajax()
+		{
+			if (!$this->app->getIdentity()->authorise('core.create', 'com_templates'))
+			{
+				// User is not authorised to delete
```
```suggestion
				// User is not authorised to create
```

dgrammatiko committed:
Update administrator/components/com_templates/src/Controller/TemplateController.php

Co-authored-by: Benjamin Trenkle <bembelimen@users.noreply.github.com>

dgrammatiko committed:
Bailout Early

dgrammatiko committed:
CS

dgrammatiko committed:
Merge branch '4.1-dev' into 4.1-dev-com_templates_clean

dgrammatiko committed:
Merge branch '4.1-dev' into 4.1-dev-com_templates_clean

RickR2H commented:
@dgrammatiko I checked out the PR but I got a some warnings in the template view.
![warnings-2](https://user-images.githubusercontent.com/5610413/142932854-3cb8fa42-a34b-474f-ba49-7a5a02c4eb96.jpg)

dgrammatiko committed:
Don't throw warnings

dgrammatiko commented:
>  @dgrammatiko I checked out the PR but I got a some warnings in the template view.

@RickR2H good catch, should be fine now

RickR2H commented:
The default thumb isn't working ;)
![Image 3](https://user-images.githubusercontent.com/5610413/142934702-8662bfc0-c98f-425b-b3a3-17c67e516031.jpg)

dgrammatiko committed:
better conditianal

RickR2H commented:
Some warnings still appear:

Warning: Undefined variable $key in C:\wamp64\www\testing\administrator\components\com_templates\tmpl\template\default.php on line 91 Call Stack #TimeMemoryFunctionLocation 10.0002366680{main}( )...\index.php:0 20.0003367448require_once( 'C:\wamp64\www\testing\administrator\includes\app.php )...\index.php:32 30.01661276768Joomla\CMS\Application\CMSApplication->execute( )...\app.php:63 40.01711359128Joomla\CMS\Application\AdministratorApplication->doExecute( )...\CMSApplication.php:278 50.07222317544Joomla\CMS\Application\AdministratorApplication->dispatch( $component = ??? )...\AdministratorApplication.php:186 60.07732448944Joomla\CMS\Component\ComponentHelper::renderComponent( $option = 'com_templates', $params = ??? )...\AdministratorApplication.php:143 70.08122511504Joomla\CMS\Dispatcher\ComponentDispatcher->dispatch( )...\ComponentHelper.php:389 80.08182522928Joomla\CMS\MVC\Controller\BaseController->execute( $task = 'display' )...\ComponentDispatcher.php:146 90.08182522928Joomla\Component\Templates\Administrator\Controller\DisplayController->display( $cachable = ???, $urlparams = ??? )...\BaseController.php:730 100.08192522928Joomla\CMS\MVC\Controller\BaseController->display( $cachable = ???, $urlparams = ??? )...\DisplayController.php:67 110.08372620904Joomla\Component\Templates\Administrator\View\Template\HtmlView->display( $tpl = ??? )...\BaseController.php:692 120.12822878352Joomla\CMS\MVC\View\HtmlView->display( $tpl = NULL )...\HtmlView.php:230 130.12842880096Joomla\CMS\MVC\View\HtmlView->loadTemplate( $tpl = NULL )...\HtmlView.php:221 140.12962901864include( 'C:\wamp64\www\testing\administrator\components\com_templates\tmpl\template\default.php )...\HtmlView.php:432 " href=""> 

Warning: Undefined property: Joomla\CMS\Object\CMSObject::$inheritable in C:\wamp64\www\testing\administrator\components\com_templates\tmpl\template\default.php on line 379

dgrammatiko committed:
More notices

RickR2H commented:
full path is added to the folder structure in the explorer on the left.
![Image-1](https://user-images.githubusercontent.com/5610413/142936583-f0565f35-fbdc-4761-bb86-6c6dd0a6e105.jpg)

dgrammatiko committed:
Fix the thumb paths

dgrammatiko commented:
> full path is added to the folder structure in the explorer on the left.

That's a winDOS thing. I have to check where that FS fails

RickR2H commented:
> > full path is added to the folder structure in the explorer on the left.
> 
> That's a winDOS thing. I have to check where that FS fails

That explains why the links are all failing. Clicking the files in the tree results in a "File not found" error and a lot of warnings. Let me know if you need more info. We can use Glip if you want, so we don't spam the Git stream to much.

Keep me posted!

dgrammatiko committed:
Fix Fat FS lookups

dgrammatiko commented:
Did https://github.com/joomla/joomla-cms/pull/35998/commits/411a9f57f97282cffb161be0fa0b52b9e852fc8c fixed the full path? Also you should have proper thumbs now

RickR2H commented:
Folders and files work. One thing I noticed is that the folders now haven an additional "/"

![image](https://user-images.githubusercontent.com/5610413/142938862-1d74c055-7b98-414c-80b6-8e2df2f0eadf.png)

RickR2H commented:
Installing the Cassy template worked. Creating the child worked too. On opening the child I get notices:

![image](https://user-images.githubusercontent.com/5610413/142939585-ce3b6b5a-5878-46b6-9cfa-c75cb52b9528.png)

RickR2H commented:
Is it correct that a child template folder only has a templateDetails.xml to start with?

dgrammatiko committed:
Fix the missing media folder for the child template

dgrammatiko commented:
> On opening the child I get notices:

Should be ok now. You will need to manually uninstall the child and recreate it

dgrammatiko committed:
don't echo the first slash on win OS

dgrammatiko committed:
.html is not in the allowed list. WOW

PhilETaylor commented on administrator/components/com_templates/src/Controller/TemplateController.php:
```diff
@@ -980,4 +984,105 @@ public function ajax()
+	/**
+	 * Method for creating a child template.
+	 *
+	 * @return  boolean     true on success, false otherwise
```
```suggestion
	 * @return  boolean   true on success, false otherwise
```

PhilETaylor commented on administrator/components/com_templates/src/Controller/TemplateController.php:
```diff
@@ -980,4 +984,105 @@ public function ajax()
+		$file       = (string) $this->input->get('file', '', 'cmd');
+
+		// Access check.
+		if (!$this->allowEdit())
```
Access check should come right after checkToken - should be the first things done

dgrammatiko replied:
done

PhilETaylor commented on administrator/components/com_templates/src/Controller/TemplateController.php:
```diff
@@ -980,4 +984,105 @@ public function ajax()
+		// Check that from name does exist and get the folder name
+		$fromName = $model->getFromName();
+
+		if (!$fromName) {
+			$this->setMessage(Text::_('COM_TEMPLATES_ERROR_INVALID_FROM_NAME'), 'error');
+
+			return false;
+		}
```
All of the `if` statements need opening brace on their own line

```suggestion
		if (!$fromName) 
		{
			$this->setMessage(Text::_('COM_TEMPLATES_ERROR_INVALID_FROM_NAME'), 'error');

			return false;
		}
```

PhilETaylor commented on administrator/components/com_templates/src/Model/TemplateModel.php:
```diff
@@ -983,12 +1005,15 @@ public function save($data)
+		$filePath = Path::clean($fileName);
 
-		// Include the extension plugins for the save events.
+			// Include the extension plugins for the save events.
```
I think this is a mistake indentation?

dgrammatiko replied:
yes, fixed

PhilETaylor commented on administrator/components/com_templates/src/Model/TemplateModel.php:
```diff
@@ -1985,4 +1989,161 @@ protected function checkFormat($ext)
+	 */
+	public function getMediaFiles()
+	{
+		$result = array();
```
Its 2021.
```suggestion
		$result = [];
```

PhilETaylor commented on administrator/components/com_templates/src/Model/TemplateModel.php:
```diff
@@ -1985,4 +1989,161 @@ protected function checkFormat($ext)
+
+		if (!isset($template->xmldata->inheritable) || (isset($template->xmldata->parent) && $template->xmldata->parent === ''))
+		{
+			return [];
```
reuse `$result` which is a blank array already.

```suggestion
			return $result;
```

PhilETaylor commented on administrator/components/com_templates/src/Model/TemplateModel.php:
```diff
@@ -1985,4 +1989,161 @@ protected function checkFormat($ext)
+			$app->enqueueMessage(Text::_('COM_TEMPLATES_DIRECTORY_NOT_WRITABLE'), 'error');
+		}
+
+		if (is_dir($path)) {
+			$result = $this->getDirectoryTree($path);
+		} else {
+			return [];
+		}
+
+		return $result;
```
reuse `$result` which is a blank array already.

```suggestion
		if (is_dir($path)) {
			$result = $this->getDirectoryTree($path);
		}

		return $result;
```

PhilETaylor commented on administrator/components/com_templates/src/Model/TemplateModel.php:
```diff
@@ -1985,4 +1989,161 @@ protected function checkFormat($ext)
+	 *
+	 * @return  boolean   true if name is not used, false otherwise
+	 *
+	 * @since	__DEPLOY_VERSION__
```
```suggestion
	 * @since  __DEPLOY_VERSION__
```

PhilETaylor commented on administrator/components/com_templates/src/Model/TemplateModel.php:
```diff
@@ -1985,4 +1989,161 @@ protected function checkFormat($ext)
+				if (!is_dir(JPATH_ROOT . '/media/templates/' . ($template->client_id === 0 ? 'site' : 'administrator') . '/' . $template->element . '_' . $newName))
+				{
+					Folder::create(JPATH_ROOT . '/media/templates/' . ($template->client_id === 0 ? 'site' : 'administrator') . '/' . $template->element . '_' . $newName);
+					file_put_contents(JPATH_ROOT . '/media/templates/' . ($template->client_id === 0 ? 'site' : 'administrator') . '/' . $template->element . '_' . $newName . '/index.json', '{}');
```
You should never be using `file_put_contents` but using the `File::write` API as that clears opcache IIRC. Nothing should directly write to files except through Joomla's File/Folder API.

dgrammatiko replied:
I was trying to find what I did in the other pr which wasn’t such a hack but I also wanted Rick to be able to test the rest. I will remove this part as it’s a cheap hack and do it the right way…

dgrammatiko replied:
Fixed without hacks

PhilETaylor commented on administrator/components/com_templates/src/View/Template/HtmlView.php:
```diff
@@ -359,4 +376,44 @@ protected function folderTree($array)
+	 * @return  string
+	 *
+	 * @note    Uses recursion
+	 * @since   3.2
```
```suggestion
	 * @since   __DEPLOY_VERSION__
```

PhilETaylor commented on administrator/components/com_templates/tmpl/template/default.php:
```diff
@@ -1,4 +1,5 @@
 <?php
+
```
revert

```suggestion
```

PhilETaylor commented on administrator/components/com_templates/tmpl/template/default.php:
```diff
@@ -30,21 +32,18 @@
 
-if ($this->type == 'image')
-{
+if ($this->type == 'image') {
```
revert all IF statements in this file, the opening brace should be on a line by itself.

dgrammatiko replied:
Thanks, somehow my vscode doesn’t like the Joomla’s cs. I will fix them asap

dgrammatiko replied:
done

PhilETaylor commented on administrator/components/com_templates/tmpl/template/default_media_folders.php:
```diff
@@ -0,0 +1,34 @@
+<?php
+
```
```suggestion
```

PhilETaylor commented on administrator/components/com_templates/tmpl/template/default_modal_file_body.php:
```diff
@@ -1,4 +1,5 @@
 <?php
+
```
```suggestion
```

PhilETaylor commented on administrator/components/com_templates/tmpl/template/default_modal_folder_body.php:
```diff
@@ -1,4 +1,5 @@
 <?php
+
```
```suggestion
```

dgrammatiko committed:
Update administrator/components/com_templates/src/Model/TemplateModel.php

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

dgrammatiko committed:
Update administrator/components/com_templates/src/Controller/TemplateController.php

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

dgrammatiko committed:
Update administrator/components/com_templates/tmpl/template/default_modal_folder_body.php

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

dgrammatiko committed:
Update administrator/components/com_templates/tmpl/template/default_media_folders.php

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

dgrammatiko committed:
Update administrator/components/com_templates/tmpl/template/default.php

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

dgrammatiko committed:
Update administrator/components/com_templates/src/View/Template/HtmlView.php

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

dgrammatiko committed:
Update administrator/components/com_templates/src/Model/TemplateModel.php

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

dgrammatiko committed:
Update administrator/components/com_templates/src/Controller/TemplateController.php

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

dgrammatiko committed:
Update administrator/components/com_templates/src/Model/TemplateModel.php

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

dgrammatiko committed:
Update administrator/components/com_templates/src/Model/TemplateModel.php

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

dgrammatiko committed:
Merge branch '4.1-dev' into 4.1-dev-com_templates_clean

dgrammatiko committed:
CS, proper media folder logic

dgrammatiko committed:
CS

dgrammatiko committed:
Meh

dgrammatiko committed:
Paths...

dgrammatiko committed:
Merge branch '4.1-dev' into 4.1-dev-com_templates_clean

bembelimen committed:
Merge branch '4.1-dev' into 4.1-dev-com_templates_clean

coolcat-creations commented:
![grafik](https://user-images.githubusercontent.com/828371/143490850-f13b636b-266b-4cbd-a2ec-011a6432725f.png)

Thank you @dgrammatiko  for this nice feature!!! 

I started to test and my main concern is, that the folder structure change is quite confusing and not a nice change for a user.

I understood that this way it's more secure and I am gonna live with tha ;-). 

But the folder structure display is very confusing. It does not display the "reality"
So to make it clear where the files are flying around, it would be good to clarify to the user in the folder structure images what happens. 

Maybe something like that:

![grafik](https://user-images.githubusercontent.com/828371/143490952-3ded374a-0943-4697-9614-55669c7bb373.png)

coolcat-creations commented:
At the overrides tab, shouldn't be there also the files from the parent template to click on?

![grafik](https://user-images.githubusercontent.com/828371/143491555-2ad316f5-1a10-4e47-a77b-32e95a9388ff.png)

coolcat-creations commented:
I made an override of index.php, and afterwards I made a change in the main template index.php but the changes have not been recognized.
![grafik](https://user-images.githubusercontent.com/828371/143491692-2b3f1a5e-6341-4d6a-ad48-e7feecb13b9b.png)

coolcat-creations commented:
I created a .js file in the media js folder, clicked on delete, i got a success message but the file is still there:

![grafik](https://user-images.githubusercontent.com/828371/143492131-e30f948b-9649-4057-bb24-4b2c53593ef1.png)

coolcat-creations commented:
Template Overview:
The **Override Files** information is wrong here too (because the index.php is behind the parent version) and it would be great if the childtemplates would be little bit intended so that its better recognizeable that they are childs.
At the create child process it would be great if it takes the username in the author column and the current date. 
Additionally,  if possible, the version could be increased by every change - but I see that there are issues finding the right versioning pattern.

![grafik](https://user-images.githubusercontent.com/828371/143492897-9f827a48-c282-4e2b-b4de-f8752fb93885.png)

chmst commented:
@coolcat-creations Agree with indenting child templates. But I think layout improvements can be added also after merge. 

About the update notification: As much as I know, this function observes only the overrides in html.

dgrammatiko commented:
> But the folder structure display is very confusing. It does not display the "reality"

The reason I didn't wrote some extensive testing instructions on these PRs was to get such valuable responses. If I had verbally explained the UI you might have thought "ah, that makes sense" but that would be cheating. Thanks! Will add the paths (easy one)

> I made an override of index.php, and afterwards I made a change in the main template index.php but the changes have not been recognized.

That's normal, the update files work only on layout overrides. The logic hasn't changed but the question is should it also include the parent template files? (btw I don't know if that is even possible, templates could have any filename as their entry point: index, component, error, 404, dimitris...) @bembelimen any input here?

> I created a .js file in the media js folder, clicked on delete, i got a success message but the file is still there:

Ok, that's a bug

> At the overrides tab, shouldn't be there also the files from the parent template to click on?

Good question. Fwiw I'm still in favour of https://user-images.githubusercontent.com/3889375/132101328-be1643c8-e9af-4b75-8645-a9162d554214.mp4 and the https://github.com/joomla/joomla-cms/pull/32896 in general but that won't happen in 4.x

> At the create child process it would be great if it takes the username in the author column and the current date.

Easily doable

> if the childtemplates would be little bit intended so that its better recognizeable that they are childs.

I'm not sure how this would look like. The infos are in a table cell

Please keep reporting anything that you feel it's not good, so I could try to fix it. 👍

chmst commented:
> I'm not sure how this would look like. The infos are in a table cell

Maybe a fa-angle-right before can do it  (in ltr languages).

dgrammatiko committed:
better tree entry names

dgrammatiko commented:
> Maybe a fa-angle-right before can do it (in ltr languages).

What about rtl? Will `fa-angle-left` be ok?

chmst commented:
@dgrammatiko  I think so. I will need a test with ltr language. No clue atm how the table looks like in rtl

dgrammatiko committed:
Use current user name, current date for child

dgrammatiko committed:
Bug, file delete

coolcat-creations commented:
Wow, thank you @dgrammatiko for the fast response and fixes!

For:
> > I made an override of index.php, and afterwards I made a change in the main template index.php but the changes have not been recognized.
> 
> That's normal, the update files work only on layout overrides. The logic hasn't changed but the question is should it also include the parent template files? (btw I don't know if that is even possible, templates could have any filename as their entry point: index, component, error, 404, dimitris...) @bembelimen any input here?

For me a childtemplate is basically an override if not it should be clarified to the user. Comparing between the parent template and the childtemplate is very very important I think... 

> Please keep reporting anything that you feel it's not good, so I could try to fix it. 👍

Thank you! I will test tomorrow again :-)

dgrammatiko committed:
Fix upload

coolcat-creations commented:
@bembelimen any feedback about the override-check issue? I think it's one very important thing to have here - its basically the first questions someone has when doing changes on templates "How to recognize changes that have been done to the main template by updates"

dgrammatiko committed:
Add a visual cue for parent/child

dgrammatiko committed:
Merge branch '4.1-dev' into 4.1-dev-com_templates_clean

dgrammatiko committed:
Add buttons to switch between list views

@bembelimen's idea

bembelimen commented:
> @bembelimen any feedback about the override-check issue? I think it's one very important thing to have here - its basically the first questions someone has when doing changes on templates "How to recognize changes that have been done to the main template by updates"

Yes, but not in this PR please. Otherwise we kill ourself (mostly @dgrammatiko ) and never get it done in a first version.
It would be a very nice feature, but is not a "must have or we'll all die" thing.

dgrammatiko committed:
Wrong comments

bembelimen commented:
> That's normal, the update files work only on layout overrides. The logic hasn't changed but the question is should it also include the parent template files? (btw I don't know if that is even possible, templates could have any filename as their entry point: index, component, error, 404, dimitris...) @bembelimen any input here?

Yes, it should be possible, but also not in this PR. I think that needs some more thoughts, like see the parent files in a different color and when you edit them in the child, the changes will be copied to the child.

I also see a compare function in the future....but not now.

chmst commented:
@dgrammatiko  a agle-left is ok

![image](https://user-images.githubusercontent.com/1035262/143572956-c25227c0-bc6e-406c-a418-ebd9bce5ce1a.png)

dgrammatiko commented:
@chmst I added a badge so the view is more like
![Screenshot 2021-11-26 at 12 19 51](https://user-images.githubusercontent.com/3889375/143573356-aed98f90-9560-43f9-8512-8b2335bb45d6.png)
Should I also add the angle icon to signify the child template?

chmst commented:
> Should I also add the angle icon to signify the child template?

Not for me

dgrammatiko committed:
Clean up the child method

dgrammatiko committed:
Merge branch '4.1-dev' into 4.1-dev-com_templates_clean

coolcat-creations commented:
the badge is nice! Like it :-)

Quy commented on administrator/components/com_templates/tmpl/template/default_modal_file_body.php:
```diff
@@ -22,7 +23,26 @@
-					<?php echo $this->loadTemplate('folders'); ?>
+					<ul class="directory-tree treeselect root">
+						<li class="folder-select">
+							<a class="folder-url" data-id="<?php echo base64_encode($key); ?>" href="" data-base="template">
```
> PHP Notice:  Undefined variable: key

dgrammatiko replied:
Fixed with 0f35de4

Quy commented:
![35998-templates](https://user-images.githubusercontent.com/368084/143665607-7fc8ba57-e2ba-4f00-9e5d-9793fddd7db6.jpg)

Please match the button icon to the quick icon.

![35998](https://user-images.githubusercontent.com/368084/143665591-c7762c64-0861-40d7-ae28-bad88b0b6cad.jpg)

brianteeman commented:
@quy the icon is unrelated to this pr. It was _wrong_ before

dgrammatiko committed:
PHP notice, use entry dir as the starting point

dgrammatiko committed:
Correct icon

bembelimen committed:
Merge branch '4.1-dev' into 4.1-dev-com_templates_clean

richard67 commented on administrator/components/com_templates/src/Controller/TemplateController.php:
```diff
@@ -74,7 +74,8 @@ public function close()
 		$id = (int) $this->input->get('id', 0, 'int');
-		$url  = 'index.php?option=com_templates&view=template&id=' . $id . '&file=' . $file;
+		$url  = 'index.php?option=com_templates&view=template&id=' . $id . '&file=' .
+		$file . '&isMedia=' . $isMedia;
```
Where is the `$isMedia` defined and set?

dgrammatiko replied:
Good catch

richard67 commented on administrator/components/com_templates/src/Controller/TemplateController.php:
```diff
@@ -131,7 +132,8 @@ public function publish()
 
-		$url  = 'index.php?option=com_templates&view=template&id=' . $id . '&file=' . $file;
+		$url  = 'index.php?option=com_templates&view=template&id=' . $id . '&file=' .
+		$file . '&isMedia=' . $isMedia;
```
Where is the `$isMedia` defined and set?

dgrammatiko replied:
Do a quick search for `isMedia` in the `administrator/components/com_templates/tmpl/template`. There are many forms to point them here. Also in a couple cases (modals iirc) this is set using JS

dgrammatiko replied:
Obviously, I misunderstood the comments...
Good catch

dgrammatiko committed:
Undefined $isMedia Pfff

richard67 commented on administrator/components/com_templates/src/Controller/TemplateController.php:
```diff
@@ -131,7 +132,8 @@ public function publish()
 
-		$url  = 'index.php?option=com_templates&view=template&id=' . $id . '&file=' . $file;
+		$url  = 'index.php?option=com_templates&view=template&id=' . $id . '&file=' .
+		$file . '&isMedia=' . $this->input->getInt('isMedia', 0);
```
```suggestion
			$file . '&isMedia=' . $this->input->getInt('isMedia', 0);
```

richard67 commented on administrator/components/com_templates/src/Controller/TemplateController.php:
```diff
@@ -74,7 +74,8 @@ public function close()
 		$id = (int) $this->input->get('id', 0, 'int');
-		$url  = 'index.php?option=com_templates&view=template&id=' . $id . '&file=' . $file;
+		$url  = 'index.php?option=com_templates&view=template&id=' . $id . '&file=' .
+		$file . '&isMedia=' . $this->input->getInt('isMedia', 0);
```
```suggestion
			$file . '&isMedia=' . $this->input->getInt('isMedia', 0);
```

dgrammatiko committed:
Update administrator/components/com_templates/src/Controller/TemplateController.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

dgrammatiko committed:
Update administrator/components/com_templates/src/Controller/TemplateController.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

dgrammatiko committed:
Merge #36011

chmst commented:
Updated a fresh 4.1 with the update file and installed cassy. 
When editing a file, the file name should be visible. Here I have the number "226" for all files: 
![image](https://user-images.githubusercontent.com/1035262/143683610-c3ae538e-de00-4c8b-bae9-c74865adcb04.png)

dgrammatiko committed:
Remove a debug line

dgrammatiko commented:
> Here I have the number "226" for all files

Fixed with https://github.com/joomla/joomla-cms/pull/35998/commits/b962afbb44aebf032bb59a97f596966b128f0e9a

dgrammatiko committed:
Windows use / instead of \

dgrammatiko commented:
@chmst could you try https://github.com/joomla/joomla-cms/pull/35998/commits/bfe1eb6b4b2cce7682314a47d9e573ddb25c3447 It should use `/` for all the directory seperators. Unfortunatelly this affects windows users and I can't test it here

chmst commented:
![image](https://user-images.githubusercontent.com/1035262/143685719-535f3d03-c9dc-4d71-a6a7-67eee487164b.png)

Ok for the slashes. 

It remains the "226" instead filename.

brianteeman commented:
> Ok for the slashes.

There is an extra slash in the editing file part

chmst commented:
I need new glasses

dgrammatiko committed:
Winows double //

dgrammatiko commented:
@chmst did https://github.com/joomla/joomla-cms/pull/35998/commits/4e76c7b4d2ab161f8efb9cf6f6544c13f32b2e02 fixed the double `//`?

chmst commented:
Actual result. 
![image](https://user-images.githubusercontent.com/1035262/143687682-0aa33e1d-1221-47ca-b797-64825f21763f.png)

dgrammatiko commented:
@chmst is there anything left here that's not fixed?

brianteeman commented on administrator/language/en-GB/com_templates.ini:
```diff
@@ -24,6 +24,7 @@ COM_TEMPLATES_BUTTON_RENAME_FILE="Rename File"
 COM_TEMPLATES_BUTTON_RESIZE="Resize"
 COM_TEMPLATES_BUTTON_UPLOAD="Upload"
 COM_TEMPLATES_CHECK_FILE_OWNERSHIP="Check file ownership"
+COM_TEMPLATES_CHILD_SUCCESS="Child template succesfully created."
```
kill the successfully - it is redundant (and spelt wrong)

```suggestion
COM_TEMPLATES_CHILD_SUCCESS="Child template created."
```

dgrammatiko committed:
Update administrator/language/en-GB/com_templates.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented on administrator/language/en-GB/com_templates.ini:
```diff
@@ -215,13 +217,20 @@ COM_TEMPLATES_TAB_DESCRIPTION="Template Description"
 COM_TEMPLATES_TEMPLATE_FILENAME="Editing file &quot;%s&quot; in template &quot;%s&quot;."
 COM_TEMPLATES_TEMPLATE_FILES="Template Files"
+COM_TEMPLATES_TEMPLATE_FOLDER="Template Folder"
+COM_TEMPLATES_TEMPLATE_IS_CHILD_OF="Child of <b class="text-capitalize text-light">%s</b>"
```
b ?

dgrammatiko replied:
bold?

brianteeman replied:
Not any more -its now called the "bring attention to" element. You probably should use strong as we do everywhere else.

dgrammatiko replied:
Ok done

chmst commented:
I found double slashes also here:
![image](https://user-images.githubusercontent.com/1035262/143688067-16100b2a-7d8f-44d6-a10f-c4d62e761859.png)

dgrammatiko committed:
double //

dgrammatiko committed:
b->strong

dgrammatiko commented:
>  found double slashes also here:

Should be fixed with https://github.com/joomla/joomla-cms/pull/35998/commits/2359e7f1cf53d9ccc6db6aa3166d8d91293d31d3

chmst commented:
Same in rename file modal ;)

richard67 commented:
@dgrammatiko Do the testing instructions already include the changes from merging in #36011 ? Or is there something missing?

dgrammatiko committed:
rename modatl tittle //

dgrammatiko commented:
> Or is there something missing?

No, the instructions need update. I'm on it

chmst commented:
Slashes in Foldernames seem to be ok now.

Quy commented on administrator/components/com_templates/tmpl/template/default_tree_media.php:
```diff
@@ -0,0 +1,73 @@
+		<?php endif; ?>
+		<?php if (is_object($value)) : ?>
+			<li>
+				<a class="file" href='<?php echo Route::_('index.php?option=com_templates&view=template&id=' . $this->id . '&file=' . $value->id . '&isMedia=1'); ?>'>
```
```suggestion
				<a class="file" href="<?php echo Route::_('index.php?option=com_templates&view=template&id=' . $this->id . '&file=' . $value->id . '&isMedia=1'); ?>">
```

dgrammatiko committed:
Update administrator/components/com_templates/tmpl/template/default_tree_media.php

Co-authored-by: Quy <quy@fluxbb.org>

Quy commented on plugins/editors/tinymce/tinymce.php:
```diff
@@ -1159,4 +1106,138 @@ private function removeElementWithValue($array, $key, $value)
+			$minifiedPath    = $path;
+			$nonMinifiedPath = substr_replace($path, '', $position, 4);
+
+			if (JDEBUG && is_file($nonMinifiedPath))
+			{
+				return Uri::root(true) . str_replace(JPATH_ROOT, '', $nonMinifiedPath);
+			}
+			elseif (is_file($minifiedPath))
+			{
+				return Uri::root(true) . str_replace(JPATH_ROOT, '', $minifiedPath);
+			}
+			else
+			{
+				return '';
+			}
+		}
+
+		$minifiedPath = pathinfo($path, PATHINFO_DIRNAME) . '/' . pathinfo($path, PATHINFO_FILENAME) . '.min.' . pathinfo($path, PATHINFO_EXTENSION);
+
+		if (JDEBUG && is_file($path))
+		{
+			return Uri::root(true) . str_replace(JPATH_ROOT, '', $path);
+		}
+		elseif (is_file($minifiedPath))
+		{
+			return Uri::root(true) . str_replace(JPATH_ROOT, '', $minifiedPath);
+		}
+		else
+		{
+			return '';
+		}
```
```suggestion
			if (JDEBUG && is_file($nonMinifiedPath))
			{
				return Uri::root(true) . str_replace(JPATH_ROOT, '', $nonMinifiedPath);
			}

			if (is_file($minifiedPath))
			{
				return Uri::root(true) . str_replace(JPATH_ROOT, '', $minifiedPath);
			}

			return '';
		}

		$minifiedPath = pathinfo($path, PATHINFO_DIRNAME) . '/' . pathinfo($path, PATHINFO_FILENAME) . '.min.' . pathinfo($path, PATHINFO_EXTENSION);

		if (JDEBUG && is_file($path))
		{
			return Uri::root(true) . str_replace(JPATH_ROOT, '', $path);
		}

		if (is_file($minifiedPath))
		{
			return Uri::root(true) . str_replace(JPATH_ROOT, '', $minifiedPath);
		}

		return '';
```

dgrammatiko committed:
Update plugins/editors/tinymce/tinymce.php

Co-authored-by: Quy <quy@fluxbb.org>

dgrammatiko committed:
Switch the fallback icon

This has a Creative Commons Zero v1.0 Universal (CC0) license
and it was copied from: https://iconduck.com/search?query=xcode

Quy commented on administrator/language/en-GB/com_templates.ini:
```diff
@@ -215,13 +217,20 @@ COM_TEMPLATES_TAB_DESCRIPTION="Template Description"
 COM_TEMPLATES_TAB_OVERRIDES="Create Overrides"
 COM_TEMPLATES_TAB_UPDATED_FILES="Updated Files"
+COM_TEMPLATES_TEMPLATE_CHILD="Child Template"
+COM_TEMPLATES_TEMPLATE_CHILD_NAME_LABEL="Name of the child template"
```
```suggestion
COM_TEMPLATES_TEMPLATE_CHILD_NAME_LABEL="Child Template Name"
```

Quy commented on administrator/components/com_templates/src/View/Styles/HtmlView.php:
```diff
@@ -114,7 +114,11 @@ public function display($tpl = null)
+		ToolbarHelper::link('index.php?option=com_templates&view=templates&client_id=' . $clientId, Text::_('COM_TEMPLATES_MANAGER_TEMPLATES'), 'icon-code thememanager');
 
 		// Set the title.
 		if ((int) $this->get('State')->get('client_id') === 1)
```
```suggestion
		if ($clientId === 1)
```

Quy commented on plugins/editors/tinymce/src/Field/TemplateslistField.php:
```diff
@@ -0,0 +1,98 @@
+   */
+  public function setup(\SimpleXMLElement $element, $value, $group = null)
+  {
+    $return = parent::setup($element, $value, $group);
```
Please use tab.

Quy commented on plugins/editors/tinymce/tinymce.php:
```diff
@@ -1159,4 +1106,136 @@ private function removeElementWithValue($array, $key, $value)
+	 */
+	protected function includeRelativeFiles($folder, $file)
+	{
+		$fallback  = Uri::root(true) . '/media/system/css/editor' . (JDEBUG ? '' : '.min') . '.css';
+		$template  = $this->getActiveSiteTemplate();
```
```suggestion
		$fallback = Uri::root(true) . '/media/system/css/editor' . (JDEBUG ? '' : '.min') . '.css';
		$template = $this->getActiveSiteTemplate();
```

Quy commented on administrator/language/en-GB/com_templates.ini:
```diff
@@ -153,6 +154,7 @@ COM_TEMPLATES_MANAGE_FOLDERS="Manage Folders"
 COM_TEMPLATES_MANAGER_ADD_STYLE="Templates: Add Style"
 COM_TEMPLATES_MANAGER_EDIT_STYLE="Templates: Edit Style"
 COM_TEMPLATES_MANAGER_STYLES="Templates: Styles"
+COM_TEMPLATES_MANAGER_STYLES_BUTTON="Styles"
 COM_TEMPLATES_MANAGER_STYLES_ADMIN="Templates: Styles (Administrator)"
```
```suggestion
COM_TEMPLATES_MANAGER_STYLES_ADMIN="Templates: Styles (Administrator)"
COM_TEMPLATES_MANAGER_STYLES_BUTTON="Styles"
```

dgrammatiko committed:
Update administrator/components/com_templates/src/View/Styles/HtmlView.php

Co-authored-by: Quy <quy@fluxbb.org>

dgrammatiko committed:
Update administrator/language/en-GB/com_templates.ini

Co-authored-by: Quy <quy@fluxbb.org>

dgrammatiko committed:
Update plugins/editors/tinymce/tinymce.php

Co-authored-by: Quy <quy@fluxbb.org>

dgrammatiko committed:
Update administrator/language/en-GB/com_templates.ini

Co-authored-by: Quy <quy@fluxbb.org>

dgrammatiko committed:
tabs

richard67 commented:
@dgrammatiko I have created a child template of the Cassy which I had installed before on a 4.1-dev with this PR applied. After that, each time when I edit and save parameters of the template style of that child, I get 3 PHP notices in my PHP error log:

```
Notice:  Trying to get property 'parent' of non-object in /joomla-cms-4.1-dev/administrator/components/com_templates/src/Model/StyleModel.php on line 422, referer: https://www.joomla-41-dev.vmubu01.vmnet2.local/administrator/index.php?option=com_templates&view=style&layout=edit&id=14
PHP Notice:  Trying to get property 'parent' of non-object in /joomla-cms-4.1-dev/administrator/components/com_templates/src/Model/StyleModel.php on line 423, referer: https://www.joomla-41-dev.vmubu01.vmnet2.local/administrator/index.php?option=com_templates&view=style&layout=edit&id=14
PHP Notice:  Trying to get property 'parent' of non-object in/joomla-cms-4.1-dev/administrator/components/com_templates/src/Model/StyleModel.php on line 423, referer: https://www.joomla-41-dev.vmubu01.vmnet2.local/administrator/index.php?option=com_templates&view=style&layout=edit&id=14
```

Could you check that?

Besides this, things seem to work ok so far.

richard67 commented:
P.S.: The PHP notices reported above happen only when saving the parameters of a child template's template style (Cassy in this case). They don't happen when editing the parameters of a normal (or copy of a normal) template's template style.

dgrammatiko commented:
@richard67 what's the data in the ``#__template_styles`? Are the parent columns filled for the child?

dgrammatiko committed:
Lang failsafe

richard67 commented:
![2021-11-28_pr-35998](https://user-images.githubusercontent.com/7413183/143785471-3e2a8821-5f2e-4cc0-bf2d-b2b4fb6fa517.png)

@dgrammatiko Looks ok to me. But can it be that more than one template style for a child template is not supported?

When I make the template style of the child template be the default, I get no CSS from the parent loaded.

![2021-11-28_pr-35998_2](https://user-images.githubusercontent.com/7413183/143785555-93e56d01-c583-4371-bfce-202dc470ff38.png)

Backend template styles:

![2021-11-28_pr-35998_3](https://user-images.githubusercontent.com/7413183/143785568-d0094419-366a-4b5a-8d16-652819bc14b6.png)

dgrammatiko commented:
The names for your child templates are off, should be `cassy_child*` instead of `cassy_cassy*` Althought this shouldn't be the problem.

> But can it be that more than one template style for a child template is not supported?

No, styles are just the JSON blob from the db so there shouldn't be a limit on the copies...

> When I make the template style of the child template be the default, I get no CSS from the parent loaded.

Does the folders `media/templates/site/cassy/...` exist? This is weird...

richard67 commented:
> Does the folders `media/templates/site/cassy/...` exist? This is weird...

@dgrammatiko Yes, it exists. I've created another child with a more suitable name but that did not help.

Another thing I have noticed is that I can created new child templates, but I cannot delete them. Is that by design or will the delete come later with another PR?

richard67 commented:
@dgrammatiko It seems it loads the template css files from the media/templates/site/cassy/css folder, but stuff from sub-folders "global", "system" and "vendor" is not included.

richard67 commented:
> @dgrammatiko It seems it loads the template css files from the media/templates/site/cassy/css folder, but stuff from sub-folders "global", "system" and "vendor" is not included.

P.S. Could that be an issue of the Cassy template?

dgrammatiko commented:
> Another thing I have noticed is that I can created new child templates, but I cannot delete them. Is that by design or will the delete come later with another PR?

Actually, you can already, got to System->Extentions, select templates in the filters and then delete any child template (probably not the most intuitive workflow)...

By the way, can you remove all the preexisted templates and try again your tests, something seems awfully wrong but I can't pinpoint what it is...

richard67 commented:
@dgrammatiko I've tried again from scratch with a new installation of this branch. This time only one language, no template style copies, nothing special. I only installed the Cassy template and made it's template style the default. That worked, i.e. frontend looked ok. Then I have created a child template named "Child_1", so the generated name at the end was "Cassy_cild_1"- That worked. But then I make the template style of this child be the default, the frontpage looks like in my screenshot, i.e. some but not all CSS seems not to be loaded.

richard67 commented:
@dgrammatiko Is it possible to test with both PRs #35874 and this here together? Or do they conflict somehow?

dgrammatiko commented:
> ut then I make the template style of this child be the default, the frontpage looks like in my screenshot, i.e. some but not all CSS seems not to be loaded.

Yeah there was something off in the cassy template, try this [Cassy_new.zip](https://github.com/joomla/joomla-cms/files/7614418/Cassy_new.zip)

> Is it possible to test with both PRs #35874 and this here together? Or do they conflict somehow?

I think there is nothing that conflicts, so they could be tested together. Remember to apply the db changes...

richard67 commented:
@dgrammatiko I have just tested now with the both PR's merged together in a branch. I don't have the probklem when creating a child template of the (now inheritable) Cassiopeia and then make the template style of that child be the default.

So it either is an issue with Cassy or it needs both PR's together.

richard67 committed:
Merge branch '4.1-dev' into 4.1-dev-com_templates_clean

richard67 commented:
I have tested this item :white_check_mark: successfully on 7a95857e955f7debeee31ef58b0b43d13e85e00f<br /><br />Template manager works well with inheritable and not inheritable templates and with child templates of inheritable templates.

The right editor.css is included when editing with TinyMCE in backend, regardless if the default template style is one of an inheritable or a not inheritable template or the child template of an inheritable template.

I've tested this PR alone with the core templates being not inheritable and the inheritable Cassy template installed (fixed version), and I also have texted it in combination with PR #35874 so the core templates were inheritable.

Both worked.

If there are minor glitches which I've missed in my tests, I think they can be fixed later.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35998">issues.joomla.org/tracker/joomla-cms/35998</a>.</sub>

chmst commented:
I have tested this item :white_check_mark: successfully on 7a95857e955f7debeee31ef58b0b43d13e85e00f<br /><br />Did all what I can imagine for child template and using templates and template styles - all went well. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35998">issues.joomla.org/tracker/joomla-cms/35998</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35998">issues.joomla.org/tracker/joomla-cms/35998</a>.</sub>

bembelimen committed:
Merge branch '4.1-dev' into 4.1-dev-com_templates_clean

Quy commented on administrator/components/com_templates/tmpl/template/default_folders.php:
```diff
@@ -8,16 +8,20 @@
  */
 
 defined('_JEXEC') or die;
+
+use Joomla\CMS\Language\Text;
+
 ksort($this->files, SORT_STRING);
```
```suggestion
defined('_JEXEC') or die;

ksort($this->files, SORT_STRING);
```

richard67 replied:
@dgrammatiko Applying @Quy 's suggestion here will not change RTC status. I confirm that the "use" is obsolete and will add back test results in the tracker when applied.

dgrammatiko committed:
Update administrator/components/com_templates/tmpl/template/default_folders.php

Co-authored-by: Quy <quy@fluxbb.org>

richard67 commented:
I've added back the previous test results in the issue tracker since the last commit was just the removal of an obsolete "use" statement.

Quy commented on plugins/editors/tinymce/tinymce.php:
```diff
@@ -305,39 +276,15 @@ public function onDisplay(
-				// Issue warning notice if the file is not found (but pass name to $content_css anyway to avoid TinyMCE error
-				if (!file_exists($templates_path . '/' . $template . '/css/' . $content_css_custom))
-				{
-					$msg = sprintf(Text::_('PLG_TINY_ERR_CUSTOMCSSFILENOTPRESENT'), $content_css_custom);
```
Remove language string?

bembelimen commented:
Thank you!
